File: go/v4/throttle.go
package ccxt

import (
	"context"
	"sync"
	"time"
)

// CustomThrottler interface for custom throttler implementations
type CustomThrottler interface {
	Throttle(ctx context.Context, cost float64) error
}

// Throttler is the default token bucket throttler implementation
type Throttler struct {
	config map[string]interface{}
	queue  []throttleRequest
	mutex  sync.Mutex
}

type throttleRequest struct {
	cost     float64
	resolver chan struct{}
}

// NewThrottler creates a new throttler with the given configuration
func NewThrottler(config map[string]interface{}) *Throttler {
	defaultConfig := map[string]interface{}{
		"refillRate":  1.0,
		"delay":       0.001,
		"capacity":    1.0,
		"maxCapacity": 2000.0,
		"tokens":      0.0,
		"cost":        1.0,
	}

	// Merge with provided config
	for k, v := range config {
		defaultConfig[k] = v
	}

	return &Throttler{
		config: defaultConfig,
		queue:  make([]throttleRequest, 0),
	}
}

// Throttle implements the CustomThrottler interface
func (t *Throttler) Throttle(ctx context.Context, cost float64) error {
	if cost == 0 {
		cost = t.config["cost"].(float64)
	}

	// Simple implementation for Go - just sleep based on cost
	delay := time.Duration(cost*1000) * time.Millisecond
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(delay):
		return nil
	}
}

// SimpleDelayThrottler implements a simple delay-based throttler
type SimpleDelayThrottler struct {
	delay time.Duration
}

// NewSimpleDelayThrottler creates a new simple delay throttler
func NewSimpleDelayThrottler(delayMs int) *SimpleDelayThrottler {
	return &SimpleDelayThrottler{
		delay: time.Duration(delayMs) * time.Millisecond,
	}
}

// Throttle implements the CustomThrottler interface
func (s *SimpleDelayThrottler) Throttle(ctx context.Context, cost float64) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(s.delay):
		return nil
	}
}

// CustomTokenBucketThrottler implements a token bucket throttler with custom parameters
type CustomTokenBucketThrottler struct {
	tokens     float64
	capacity   float64
	refillRate float64
	lastRefill time.Time
	mutex      sync.Mutex
}

// NewCustomTokenBucketThrottler creates a new token bucket throttler
func NewCustomTokenBucketThrottler(capacity, refillRate float64) *CustomTokenBucketThrottler {
	return &CustomTokenBucketThrottler{
		tokens:     capacity,
		capacity:   capacity,
		refillRate: refillRate,
		lastRefill: time.Now(),
	}
}

// Throttle implements the CustomThrottler interface
func (t *CustomTokenBucketThrottler) Throttle(ctx context.Context, cost float64) error {
	t.mutex.Lock()
	defer t.mutex.Unlock()

	// Refill tokens based on time passed
	now := time.Now()
	timePassed := now.Sub(t.lastRefill).Seconds()
	t.tokens = min(t.capacity, t.tokens+timePassed*t.refillRate)
	t.lastRefill = now

	// If not enough tokens, wait
	if t.tokens < cost {
		waitTime := (cost - t.tokens) / t.refillRate
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(time.Duration(waitTime * float64(time.Second))):
			t.tokens = 0
		}
	} else {
		t.tokens -= cost
	}

	return nil
}

// AdaptiveThrottler implements an adaptive throttler that adjusts based on response times
type AdaptiveThrottler struct {
	baseDelay     time.Duration
	maxDelay      time.Duration
	currentDelay  time.Duration
	successCount  int
	errorCount    int
	mutex         sync.Mutex
}

// NewAdaptiveThrottler creates a new adaptive throttler
func NewAdaptiveThrottler(baseDelayMs, maxDelayMs int) *AdaptiveThrottler {
	return &AdaptiveThrottler{
		baseDelay:    time.Duration(baseDelayMs) * time.Millisecond,
		maxDelay:     time.Duration(maxDelayMs) * time.Millisecond,
		currentDelay: time.Duration(baseDelayMs) * time.Millisecond,
	}
}

// Throttle implements the CustomThrottler interface
func (a *AdaptiveThrottler) Throttle(ctx context.Context, cost float64) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(a.currentDelay):
		return nil
	}
}

// OnSuccess should be called after successful requests
func (a *AdaptiveThrottler) OnSuccess() {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	a.successCount++
	if a.successCount >= 10 {
		// Reduce delay on success
		a.currentDelay = max(a.baseDelay, time.Duration(float64(a.currentDelay)*0.9))
		a.successCount = 0
	}
}

// OnError should be called after failed requests
func (a *AdaptiveThrottler) OnError() {
	a.mutex.Lock()
	defer a.mutex.Unlock()

	a.errorCount++
	if a.errorCount >= 3 {
		// Increase delay on errors
		a.currentDelay = min(a.maxDelay, time.Duration(float64(a.currentDelay)*1.5))
		a.errorCount = 0
	}
}

// ExchangeAwareThrottler implements a rate limiter that respects exchange-specific limits
type ExchangeAwareThrottler struct {
	exchangeID string
	limits     map[string]exchangeLimit
	mutex      sync.Mutex
}

type exchangeLimit struct {
	requests  int
	window    time.Duration
	lastReset time.Time
}

// NewExchangeAwareThrottler creates a new exchange-aware throttler
func NewExchangeAwareThrottler(exchangeID string) *ExchangeAwareThrottler {
	t := &ExchangeAwareThrottler{
		exchangeID: exchangeID,
		limits:     make(map[string]exchangeLimit),
	}
	t.initializeLimits()
	return t
}

func (e *ExchangeAwareThrottler) initializeLimits() {
	// Example limits for different exchanges
	exchangeLimits := map[string]struct {
		requests int
		window   time.Duration
	}{
		"binance": {requests: 1200, window: 60 * time.Second}, // 1200 requests per minute
		"coinbase": {requests: 30, window: time.Second},       // 30 requests per second
		"kraken":  {requests: 15, window: time.Second},        // 15 requests per second
	}

	if limit, exists := exchangeLimits[e.exchangeID]; exists {
		e.limits["default"] = exchangeLimit{
			requests:  limit.requests,
			window:    limit.window,
			lastReset: time.Now(),
		}
	}
}

// Throttle implements the CustomThrottler interface
func (e *ExchangeAwareThrottler) Throttle(ctx context.Context, cost float64) error {
	e.mutex.Lock()
	defer e.mutex.Unlock()

	limit, exists := e.limits["default"]
	if !exists {
		// No specific limit, use default delay
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(100 * time.Millisecond):
			return nil
		}
	}

	now := time.Now()

	// Reset counter if window has passed
	if now.Sub(limit.lastReset) > limit.window {
		limit.requests = limit.requests
		limit.lastReset = now
	}

	// If we've exceeded the limit, wait
	if limit.requests <= 0 {
		waitTime := limit.window - now.Sub(limit.lastReset)
		select {
		case <-ctx.Done():
			return ctx.Err()
		case <-time.After(waitTime):
			limit.requests = limit.requests
			limit.lastReset = time.Now()
		}
	}

	limit.requests -= int(cost)
	e.limits["default"] = limit

	return nil
}

// ExponentialBackoffThrottler implements exponential backoff throttling
type ExponentialBackoffThrottler struct {
	initialDelay   time.Duration
	maxDelay       time.Duration
	backoffFactor  float64
	currentDelay   time.Duration
	consecutiveErrors int
	mutex          sync.Mutex
}

// NewExponentialBackoffThrottler creates a new exponential backoff throttler
func NewExponentialBackoffThrottler(initialDelayMs, maxDelayMs int, backoffFactor float64) *ExponentialBackoffThrottler {
	initialDelay := time.Duration(initialDelayMs) * time.Millisecond
	return &ExponentialBackoffThrottler{
		initialDelay:  initialDelay,
		maxDelay:      time.Duration(maxDelayMs) * time.Millisecond,
		backoffFactor: backoffFactor,
		currentDelay:  initialDelay,
	}
}

// Throttle implements the CustomThrottler interface
func (e *ExponentialBackoffThrottler) Throttle(ctx context.Context, cost float64) error {
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-time.After(e.currentDelay):
		return nil
	}
}

// OnSuccess resets the delay on success
func (e *ExponentialBackoffThrottler) OnSuccess() {
	e.mutex.Lock()
	defer e.mutex.Unlock()

	e.currentDelay = e.initialDelay
	e.consecutiveErrors = 0
}

// OnError increases the delay on error with exponential backoff
func (e *ExponentialBackoffThrottler) OnError() {
	e.mutex.Lock()
	defer e.mutex.Unlock()

	e.consecutiveErrors++
	e.currentDelay = min(e.maxDelay, time.Duration(float64(e.currentDelay)*e.backoffFactor))
}

// Helper functions
func min(a, b time.Duration) time.Duration {
	if a < b {
		return a
	}
	return b
}

func max(a, b time.Duration) time.Duration {
	if a > b {
		return a
	}
	return b
}

File: python/ccxt/base/throttle.py
# -*- coding: utf-8 -*-

"""Throttler implementations for CCXT"""

import asyncio
import time
from abc import ABC, abstractmethod
from typing import Optional, Dict, Any

class CustomThrottler(ABC):
    """Abstract base class for custom throttlers"""
    
    @abstractmethod
    async def throttle(self, cost: Optional[float] = None) -> None:
        """Throttle the request based on the cost"""
        pass

class Throttler:
    """Default token bucket throttler implementation"""
    
    def __init__(self, config: Optional[Dict[str, Any]] = None):
        self.config = {
            'refillRate': 1.0,
            'delay': 0.001,
            'capacity': 1.0,
            'maxCapacity': 2000,
            'tokens': 0,
            'cost': 1.0,
        }
        if config:
            self.config.update(config)
        self.queue = []
        self.running = False

    async def loop(self):
        last_timestamp = time.time()
        while self.running:
            if not self.queue:
                self.running = False
                break
                
            resolver, cost = self.queue[0]
            if self.config['tokens'] >= 0:
                self.config['tokens'] -= cost
                resolver()
                self.queue.pop(0)
                await asyncio.sleep(0)  # context switch
                if not self.queue:
                    self.running = False
            else:
                await asyncio.sleep(self.config['delay'])
                current = time.time()
                elapsed = current - last_timestamp
                last_timestamp = current
                tokens = self.config['tokens'] + (self.config['refillRate'] * elapsed)
                self.config['tokens'] = min(tokens, self.config['capacity'])

    def throttle(self, cost: Optional[float] = None) -> asyncio.Future:
        """Throttle a request with the given cost"""
        loop = asyncio.get_event_loop()
        future = loop.create_future()
        
        if len(self.queue) > self.config['maxCapacity']:
            raise Exception(f'throttle queue is over maxCapacity ({self.config["maxCapacity"]})')
        
        cost = cost if cost is not None else self.config['cost']
        self.queue.append((future.set_result, cost))
        
        if not self.running:
            self.running = True
            asyncio.create_task(self.loop())
        
        return future

# Example custom throttler implementations

class SimpleDelayThrottler(CustomThrottler):
    """Simple delay-based throttler"""
    
    def __init__(self, delay_seconds: float = 1.0):
        self.delay_seconds = delay_seconds
    
    async def throttle(self, cost: Optional[float] = None) -> None:
        await asyncio.sleep(self.delay_seconds)

class CustomTokenBucketThrottler(CustomThrottler):
    """Custom token bucket throttler with configurable parameters"""
    
    def __init__(self, capacity: float = 10.0, refill_rate: float = 1.0):
        self.tokens = capacity
        self.capacity = capacity
        self.refill_rate = refill_rate
        self.last_refill = time.time()
    
    async def throttle(self, cost: float = 1.0) -> None:
        # Refill tokens based on time passed
        now = time.time()
        time_passed = now - self.last_refill
        self.tokens = min(self.capacity, self.tokens + time_passed * self.refill_rate)
        self.last_refill = now
        
        # If not enough tokens, wait
        if self.tokens < cost:
            wait_time = (cost - self.tokens) / self.refill_rate
            await asyncio.sleep(wait_time)
            self.tokens = 0
        else:
            self.tokens -= cost

class AdaptiveThrottler(CustomThrottler):
    """Adaptive throttler that adjusts based on response times"""
    
    def __init__(self, base_delay: float = 0.1, max_delay: float = 5.0):
        self.base_delay = base_delay
        self.max_delay = max_delay
        self.current_delay = base_delay
        self.success_count = 0
        self.error_count = 0
    
    async def throttle(self, cost: Optional[float] = None) -> None:
        await asyncio.sleep(self.current_delay)
    
    def on_success(self) -> None:
        """Call this method after successful requests"""
        self.success_count += 1
        if self.success_count >= 10:
            # Reduce delay on success
            self.current_delay = max(self.base_delay, self.current_delay * 0.9)
            self.success_count = 0
    
    def on_error(self) -> None:
        """Call this method after failed requests"""
        self.error_count += 1
        if self.error_count >= 3:
            # Increase delay on errors
            self.current_delay = min(self.max_delay, self.current_delay * 1.5)
            self.error_count = 0

class ExchangeAwareThrottler(CustomThrottler):
    """Rate limiter that respects exchange-specific limits"""
    
    def __init__(self, exchange_id: str):
        self.exchange_id = exchange_id
        self.limits = {}
        self._initialize_limits()
    
    def _initialize_limits(self) -> None:
        """Initialize exchange-specific rate limits"""
        exchange_limits = {
            'binance': {'requests': 1200, 'window': 60.0},  # 1200 requests per minute
            'coinbase': {'requests': 30, 'window': 1.0},    # 30 requests per second
            'kraken': {'requests': 15, 'window': 1.0},      # 15 requests per second
        }
        
        if self.exchange_id in exchange_limits:
            limit = exchange_limits[self.exchange_id]
            self.limits['default'] = {
                'requests': limit['requests'],
                'window': limit['window'],
                'last_reset': time.time()
            }
    
    async def throttle(self, cost: float = 1.0) -> None:
        limit = self.limits.get('default')
        if not limit:
            # No specific limit, use default delay
            await asyncio.sleep(0.1)
            return
        
        now = time.time()
        
        # Reset counter if window has passed
        if now - limit['last_reset'] > limit['window']:
            limit['requests'] = limit['requests']
            limit['last_reset'] = now
        
        # If we've exceeded the limit, wait
        if limit['requests'] <= 0:
            wait_time = limit['window'] - (now - limit['last_reset'])
            await asyncio.sleep(wait_time)
            limit['requests'] = limit['requests']
            limit['last_reset'] = time.time()
        
        limit['requests'] -= cost